### ISSUE TITLE
Cannot set compression quality when writing TIFF with JPEG compression?

### ISSUE THREAD

**Issue by gcmvy3** [haraldk/TwelveMonkeys#358]
Hi,
I am using TwelveMonkeys to read in a TIFF file, make some changes to the image, and then write it back out as a new TIFF file. The input files are all JPEG compressed (some type 6, some type 7) and I want the output files to all use type 7 JPEG compression.

The problem:
When I am creating the output files, I would like to manually set the quality of JPEG compression. I am using ImageWriteParam's setCompressionQuality() method as I would normally do, but no matter what value I pass it the compression quality always comes out the same. This is an issue since I need the files to be small, and the default compression quality makes them too big.

The relevant part of my code looks like this:

        //Create an ImageWriter with metadata taken from the ImageReader
        ImageWriter imageWriter = ImageIO.getImageWriter(imageReader);
        imageWriter.setOutput(ImageIO.createImageOutputStream(output));
        
        //Explicitly set compression mode to new-type JPEG (type 7)
        //Otherwise it will try to write in type 6 for some images and fail
        ImageWriteParam writeParam = imageWriter.getDefaultWriteParam();
        writeParam.setCompressionMode(ImageWriteParam.MODE_EXPLICIT);
        writeParam.setCompressionType("JPEG"); //This line does what it's supposed to
        writeParam.setCompressionQuality(0.7f); //This line does not seem to do anything
        
        //Write the image with the specified parameters
        imageWriter.write(null, new IIOImage(inputImage, null, iioImage.getMetadata()), writeParam);

What's interesting is that setting the compression type works, but not the compression quality. Am I doing something wrong here? Or is this a bug? Thanks in advance for your help.

EDIT: fixed a stupid mistake in the code, but the main issue still persists

**Comment by haraldk** [comment]
Hi Grant, 

Sorry, but yes you are right. The compression quality setting isn't implemented for JPEG compression. This is probably an oversight on my part, as it is implemented for other compressions where it makes sense. 

Seems pretty straight forward to implement though, when I find the time. 

If you don't mind getting your hands dirty, you could probably implement this yourself. Just find the location in TIFFImageWriter.writePage where JPEG is handled, create a new ImageWriteParam set the mode to explicit and copy the setting from the original write param (if you got one). Pass it on to the JPEGImageWriter.write method. 


Best regards,

-- 
Harald K

**Comment by BobsCandyCanes** [comment]
I looked at the TIFFImageWriter source and I now see the issue. Unfortunately I couldn't get the project to build (not an issue with the library, I'm just new to Maven and I don't know what I'm doing). I'll keep trying, thanks for your help.

P.S. - This library is great. My company had some weird type 6 JPEG-in-TIFF files that we needed to process in Java. No other libraries could read them, not even JAI. TwelveMonkeys processed them no problem. So thanks for all your hard work!
